include/rmm/mr/device/sub_memory_resource.hpp

```diff
@@ -0,0 +1,283 @@
+    block b{reinterpret_cast<char*>(p), size, true};
+    heap_blocks[b.ptr] = b;
+    return b;
+  }
+
+  void free_all()
+  {
+    for (auto b : heap_blocks) heap_resource->deallocate(b.first, b.second.size);
+    heap_blocks.clear();
+  }
+
+#ifndef NDEBUG
+  void print() {
+    std::cout << "allocated_blocks: " << allocated_blocks.size() << "\n";
+    for (auto b : allocated_blocks) { b.print(); }
+
+    std::cout << "no-sync free blocks: ";
+    no_sync_blocks.print();
+
+    std::cout << "sync free blocks: ";
+    for (auto s : sync_blocks) { 
+      std::cout << "stream " << s.first << " ";
+      s.second.print();
+    }
+    std::cout << "\n";
+  }
+#endif // DEBUG
+
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param The stream to associate this allocation with
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+    if (bytes <= 0) return nullptr;
+    bytes = rmm::detail::align_up(bytes, allocation_alignment);
+    block b = available_larger_block(bytes, stream);
+    void* p = allocate_from_block(b, bytes);
+#ifndef NDEBUG
+    //std::cout << "Allocate " << bytes << " B on stream " << stream << "\n";
+    //print();
+#endif
+    return p;
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws std::runtime_error if \p p was not allocated by this resource.
+   *
+   * @param p Pointer to be deallocated
+   *---------------------------------------------------------------------------**/
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override {
+    find_and_free_block(p, bytes, stream);
+#ifndef NDEBUG
+    //std::cout << "Free " << bytes << " B on stream " << stream 
+    //          << " pointer: " << p << "\n";
+    //print();
+#endif
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws nothing
+   *
+   * @param stream to execute on
+   * @return std::pair contaiing free_size and total_size of memory
+   *---------------------------------------------------------------------------**/
+  std::pair<size_t,size_t> do_get_mem_info( cudaStream_t stream) const override {
+    std::size_t free_size{};
+    std::size_t total_size{};
+    // TODO implement this
+    return std::make_pair(free_size, total_size);
+  }
+
+  size_t maximum_pool_size{default_maximum_size};
+
+  device_memory_resource *heap_resource;
```

Comment:
Likewise, the `heap_resource` (`Upstream`) should be a template to eliminate the extra virtual function call.